modules/http/http.c

```diff
@@ -173,46 +175,65 @@ _set_curl_opt(HTTPDestinationDriver *self, LogMessage *msg, struct curl_slist *c
   curl_easy_setopt(self->curl, CURLOPT_SSL_VERIFYHOST, self->peer_verify ? 2L : 0L);
   curl_easy_setopt(self->curl, CURLOPT_SSL_VERIFYPEER, self->peer_verify ? 1L : 0L);
 
-  curl_easy_setopt(self->curl, CURLOPT_HTTPHEADER, curl_headers);
-
-  const gchar *body = body_rendered ? body_rendered->str : log_msg_get_value(msg, LM_V_MESSAGE, NULL);
+  curl_easy_setopt(self->curl, CURLOPT_TIMEOUT, self->timeout);
 
-  curl_easy_setopt(self->curl, CURLOPT_POSTFIELDS, body);
   if (self->method_type == METHOD_TYPE_PUT)
     curl_easy_setopt(self->curl, CURLOPT_CUSTOMREQUEST, "PUT");
 }
 
+static void
+_set_payload(HTTPDestinationDriver *self, struct curl_slist *curl_headers, const gchar *body)
+{
+  curl_easy_setopt(self->curl, CURLOPT_HTTPHEADER, curl_headers);
+  curl_easy_setopt(self->curl, CURLOPT_POSTFIELDS, body);
+}
+
 static worker_insert_result_t
 _insert(LogThrDestDriver *s, LogMessage *msg)
 {
   CURLcode ret;
+  worker_insert_result_t retval;
 
   HTTPDestinationDriver *self = (HTTPDestinationDriver *) s;
 
   struct curl_slist *curl_headers = _get_curl_headers(self, msg);
-  GString *body_rendered = _get_body_rendered(self, msg);
-
-  _set_curl_opt(self, msg, curl_headers, body_rendered);
+  const gchar *body = _get_body(self, msg);
+  _set_payload(self, curl_headers, body);
 
   if ((ret = curl_easy_perform(self->curl)) != CURLE_OK)
     {
       msg_error("curl: error sending HTTP request",
                 evt_tag_str("error", curl_easy_strerror(ret)));
 
-      if (body_rendered)
-        g_string_free(body_rendered, TRUE);
-
       curl_slist_free_all(curl_headers);
 
-      return WORKER_INSERT_RESULT_ERROR;
+      return WORKER_INSERT_RESULT_NOT_CONNECTED;
     }
 
-  if (body_rendered)
-    g_string_free(body_rendered, TRUE);
+  glong http_code = 0;
+  curl_easy_getinfo (self->curl, CURLINFO_RESPONSE_CODE, &http_code);
+  switch (http_code/100)
+    {
```

Comment:
I think you should extract the error handling code to a function of its own (e.g. _map_http_status_to_worker_status)